rust_src/remacs-sys/lib.rs

```diff
@@ -341,6 +341,122 @@ pub struct CursorPos {
     vpos: c_int,
 }
 
+#[repr(C)]
+pub struct glyph {
+    /// Position from which this glyph was drawn.  If `object' below is a
+    /// Lisp string, this is an index into that string.  If it is a
+    /// buffer, this is a position in that buffer.  In addition, some
+    /// special glyphs have special values for this:
+    ///
+    /// glyph standing for newline at end of line    0
+    /// empty space after the end of the line       -1
+    /// overlay arrow on a TTY                      -1
+    /// glyph displaying line number                -1
+    /// glyph at EOB that ends in a newline         -1
+    /// left truncation glyphs:                     -1
+    /// right truncation/continuation glyphs        next buffer position
+    /// glyph standing for newline of an empty line buffer position of newline
+    /// stretch glyph at left edge of R2L lines     buffer position of newline
+    charpos: ptrdiff_t,
+
+    /// Lisp object source of this glyph.  Currently either a buffer or a
+    /// string, if the glyph was produced from characters which came from
+    /// a buffer or a string; or nil if the glyph was inserted by
+    /// redisplay for its own purposes, such as padding, truncation, or
+    /// continuation glyphs, or the overlay-arrow glyphs on TTYs.
+    object: Lisp_Object,
+
+    /// Width in pixels.
+    pixel_width: i16,
+}
+
+#[repr(C)]
+pub struct glyph_pool {
+    // Vector of glyphs allocated from the heap.
+    glyphs: *mut glyph,
+
+    // Allocated size of `glyphs'.
+    nglyphs: ptrdiff_t,
+
+    // Number of rows and columns in a matrix.
+    nrows: c_int,
+    ncolumns: c_int,
+}
+
+#[repr(C)]
+pub struct glyph_matrix {
+    pub pool: *mut glyph_pool,
+    pub rows: *mut glyph_row,
+    pub rows_allocated: ptrdiff_t,
+    pub nrows: c_int,
+    /* more to come */
+}
+
+#[repr(C)]
+pub struct glyph_row {
+    /// Pointers to beginnings of areas.  The end of an area A is found at
+    /// A + 1 in the vector.  The last element of the vector is the end
+    /// of the whole row.
+    ///
+    /// Kludge alert: Even if used[TEXT_AREA] == 0, glyphs[TEXT_AREA][0]'s
+    /// position field is used.  It is -1 if this row does not correspond
+    /// to any text; it is some buffer position if the row corresponds to
+    /// an empty display line that displays a line end.  This is what old
+    /// redisplay used to do.  (Except in code for terminal frames, this
+    /// kludge is no longer used, I believe. --gerd).
+    ///
+    /// See also start, end, displays_text_p and ends_at_zv_p for cleaner
+    /// ways to do it.  The special meaning of positions 0 and -1 will be
+    /// removed some day, so don't use it in new code.
+    glyphs: *mut glyph, // really an array
```

Comment:
you could do `[*mut glyph; 4usize]`, or bring in `LAST_SIZE`. Either way, you should declare it to be the right size, otherwise the struct won't be the right size.

Reply:
hmm, I will try again then. I had issues last time.

Reply:
Oh, I remember now. I could not find a guarantee that a Rust array would match a C array in memory.